### PR TITLE
Split sygus grammar to its own ANTLR grammar

### PR DESCRIPTION
**Pull request by ajreynol**

### CODE REVIEW

**Comment by ajreynol** [comment]
Needs clang-format.

File: src/parser/smt2/Smt2.g [hunk 1]
```diff
@@ -900,15 +797,151 @@ sygusCommand [std::unique_ptr<CVC4::Command>* cmd]
...
+/** Reads a sygus grammar
+ *
+ * The resulting sygus datatype encoding the grammar is stored in ret.
+ * The argument sygus_vars indicates the sygus bound variable list, which is
+ * the argument list of the function-to-synthesize (or null if the grammar
+ * has bound variables).
+ * The argument fun is a unique identifier to avoid naming clashes for the
+ * datatypes constructed by this call.
+ */
+sygusGrammar[CVC4::Type & ret,
```
Comment (HanielB) [inline src/parser/smt2/Smt2.g]: you can remove the `CVC4::` prefixes everywhere

File: src/parser/smt2/Smt2.g [hunk 2]
```diff
@@ -900,15 +797,151 @@ sygusCommand [std::unique_ptr<CVC4::Command>* cmd]
...
+   // but they are necessary to parse SyGuS examples
+   LPAREN_TOK(sygusGTerm[sgts.back().back(), fun] {
+     sgts.back().push_back(CVC4::SygusGTerm());
+   })
+   + RPAREN_TOK { sgts.back().pop_back(); } RPAREN_TOK)
+      + RPAREN_TOK
+  {
+    Debug("parser-sygus") << "--- Process " << sgts.size() << " sygus gterms..."
+                          << std::endl;
+    for (unsigned i = 0; i < sgts.size(); i++)
```
Comment (HanielB) [inline src/parser/smt2/Smt2.g]: cache size

File: src/parser/smt2/Smt2.g [hunk 3]
```diff
@@ -900,15 +797,151 @@ sygusCommand [std::unique_ptr<CVC4::Command>* cmd]
...
+    {
+      for (unsigned j = 0; j < sgts[i].size(); j++)
```
Comment (HanielB) [inline src/parser/smt2/Smt2.g]: ditto

File: src/parser/smt2/Smt2.g [hunk 4]
```diff
@@ -900,15 +797,151 @@ sygusCommand [std::unique_ptr<CVC4::Command>* cmd]
...
+                                        unresolved_gterm_sym,
+                                        sygus_vars,
+                                        sygus_to_builtin,
+                                        sygus_to_builtin_expr,
+                                        sub_ret);
+      }
+    }
+    // swap index if necessary
+    Debug("parser-sygus") << "--- Making sygus datatypes..." << std::endl;
+    for (unsigned i = 0; i < datatypes.size(); i++)
```
Comment (HanielB) [inline src/parser/smt2/Smt2.g]: ditto

File: src/parser/smt2/Smt2.g [hunk 5]
```diff
@@ -900,15 +797,151 @@ sygusCommand [std::unique_ptr<CVC4::Command>* cmd]
...
+    {
+      Debug("parser-sygus") << "..." << datatypes[i].getName()
+                            << " has builtin sort " << sorts[i] << std::endl;
+    }
+    Expr bvl;
+    if (!sygus_vars.empty())
+    {
+      bvl = MK_EXPR(kind::BOUND_VAR_LIST, sygus_vars);
+    }
+    for (unsigned i = 0; i < datatypes.size(); i++)
```
Comment (HanielB) [inline src/parser/smt2/Smt2.g]: ditto

File: src/parser/smt2/Smt2.g [hunk 6]
```diff
@@ -900,15 +797,151 @@ sygusCommand [std::unique_ptr<CVC4::Command>* cmd]
...
+                                    cnames[i],
+                                    cargs[i],
+                                    unresolved_gterm_sym[i],
+                                    sygus_to_builtin);
+    }
+    PARSER_STATE->setSygusStartIndex(fun, startIndex, datatypes, sorts, ops);
+    PARSER_STATE->popScope();
+    Debug("parser-sygus") << "--- Make " << datatypes.size()
+                          << " mutual datatypes..." << std::endl;
+    for (unsigned i = 0; i < datatypes.size(); i++)
```
Comment (HanielB) [inline src/parser/smt2/Smt2.g]: ditto

File: src/parser/smt2/Smt2.g (hunk 1 above)
Comment (ajreynol) [reply src/parser/smt2/Smt2.g]: Actually, the way the ANTLR C++ is generated, these CVC4:: are required.

**Pull request merged**